### ISSUE TITLE
Code review

### ISSUE THREAD

**Issue by BigDeepBlue** [emiliskan/notifications_sprint_1#31]
Отличная работа! У меня не серьезных замечаний, только некоторое число рекомендаций.
Хочу дать вам одну рекомендацию по поводу работы с гитом и github: обязательно сквошьте коммиты в PR прежде чем мержить их в мастер ветку, так история коммитов будет более линейная и читаемая. В случае совместной работы как [здесь](https://github.com/emiliskan/notifications_sprint_1/pull/27) можно было сделать два коммита, по одному от каждого из авторов.
Теперь к замечаниям по самому проекту (они в порядке чтения, а не важности):
1. https://github.com/emiliskan/notifications_sprint_1/blob/7d8f5fec797d21724ecd884d8de791e55128b441/docker-compose.yml#L48 не очень понял почему для воркера не использовали `.env` файлик, как для всех остальных сервисов
2. https://github.com/emiliskan/notifications_sprint_1/blob/7d8f5fec797d21724ecd884d8de791e55128b441/admin/Dockerfile#L10 по COPY, лучше использовать glob-синтаксис, чтобы одной строкой скопировать оба файлика, таким образом сэкономите один слой докера
3. https://github.com/emiliskan/notifications_sprint_1/blob/7d8f5fec797d21724ecd884d8de791e55128b441/admin/entrypoint.sh#L14 для разработки такая точка входа норм, в продакшене лучше все миграции и сбор статики делать через отдельный сервис, использующий этот же Dockerfile, чтобы не выполнять эти команды на каждый старт контейнера с админкой
4. Зависимости в pipfile лучше все-таки фиксировать, чтобы при добавлении новой, случайно не апнуть что-нибудь, как это недавно случилось у многих при выходе aioredis==2.0.0, где сильно поменялись интерфейсы
5. https://github.com/emiliskan/notifications_sprint_1/blob/7d8f5fec797d21724ecd884d8de791e55128b441/admin/notify/forms.py#L15 а что будете делать, когда типо станет больше разумного числа для выпадающего списка?)
6. subject из шаблона тоже было бы хорошо в истории хранить, для почтовых рассылок - это очень важное поле, было бы хорошо знать, что в нем отправилось. и вопрос туда же..почему в [форме](https://github.com/emiliskan/notifications_sprint_1/blob/main/admin/notify/forms.py) нет темы (subject)?
7. 

p.s. чуть позже докину еще

**Comment by vctecc** [comment]
1. Костыль. Celery видит джанговскую переменную окружения, но не видет приложение. И начинает об этом пистаь в лог.

**Comment by BigDeepBlue** [comment]
> Костыль.
костыли-костылики) но ок, норм

**Comment by BigDeepBlue** [comment]
> Наверное нужны уточнения к вопросу, не всё в нем понятно

я имел ввиду, что на большом числе писем в одинаковых рассылках (вышел новый фильм или что-то в этом роде) вы придете к мылси, что надо заюзать почтового-провайдера (кстати, смс-провайдеры тоже есть), вроде mailgun, чтобы не работать самостоятельно с SMTP или POP3. так как объемы будут очень большие по числу ящиков, а письма одинаковые (шаблонизацию многие сервисы тоже поддерживают). так вот у меня был вопрос к текущим интерфейсам. реализация такова, что вам придется достаточно сильно переписать код, чтобы реализовать такой функционал

**Comment by BigDeepBlue** [comment]
по остальному отвечу уже после ваших правок на втором ревью

**Comment by vctecc** [comment]
> вы придете к мылси, что надо заюзать почтового-провайдера (кстати, смс-провайдеры тоже есть), вроде mailgun, чтобы не работать самостоятельно с SMTP или POP3

Сразу отказались от использования чистого SMTP, письма ходят через провайдера Sendgrid

**Comment by vctecc** [comment]
Живая дискуссия в репозитории, это круто. Очень не хватало этого в прошлом ревью, спасибо!

**Comment by BigDeepBlue** [comment]
> Сразу отказались от использования чистого SMTP, письма ходят через провайдера Sendgrid

да, точно) подзабыл пока писал коммент) тогда вдвойне странно, что вы не предусмотрели отправку одного письма нескольким адресатам за один запрос) в этом же основная фишка всех провайдеров =)

**Comment by BigDeepBlue** [comment]
> Живая дискуссия в репозитории

тут вам свезло, что я глянул почту и увидел ваш ответ) в целом мы все-таки чисто физически не можем между ревью отвечать, процесс построен по итерациям, если будем отвечать еще между - будет просто много времени потрачено :)
Но я тоже был рад общению

**Comment by BigDeepBlue** [comment]
> Вместо копирования пробрасывается volume в docker compose файле, там же запуск

это конечно здорово, но что вы будете делать при переезде в кубер?)) придется дописывать и копирование и команду запуска

**Comment by emiliskan** [comment]
> > Вместо копирования пробрасывается volume в docker compose файле, там же запуск
> 
> это конечно здорово, но что вы будете делать при переезде в кубер?)) придется дописывать и копирование и команду запуска

я не знаю ничего про кубер( но спасибо, что подметили, у нас как раз была дискуссия в чем разница этих подходов

**Comment by BigDeepBlue** [comment]
Остальные замечания в PR:
https://github.com/emiliskan/notifications_sprint_1/pull/39
И конечно очень не хватает тестов хоть на что-то) Код без тестов поддерживать очень грустно, давайте хотя бы минимальные (юниты) напишем на часть логики.
И давайте напишем чуть более сложную логику обработки ошибок [тут](https://github.com/emiliskan/notifications_sprint_1/blob/7c74fa925ee1598ed7b0d66624ab830aec89dd44/senders/main.py#L27). Помните, что писали в учебнике, уведомления - самая большая боль, когда они работают не так. Сходу одна прикольная ситуация: письмо отправилось, но мы словили фейл от пг при записи истории в _save_history, в итоге ретрай отправит тоже самое письмо еще раз, пользователь злой, вы грустите

**Comment by BigDeepBlue** [comment]
> я не знаю ничего про кубер( 

Это то, с помощью чего с большой вероятностью ваше приложение будет деплоится в крупной компании (jenkins, ручной деплой по ssh уже уходят в прошлое)
официальная страница:
https://kubernetes.io/ru/
по традиции вводная статья на хабре:
https://habr.com/ru/post/258443/

В общем, даже без кубера, с большой вероятностью приложение будет деплоится в CI через docker, без docker-compose. И тогда у вас проблемы :)

docker-compose сейчас остается как инструмент для локального тестирования, e2e-тестов, а также его юзаются девопсы, чтобы развернуть, что-нибудь однонодовое в небольшой компании, тот же elk-стэк, например

**Comment by vctecc** [comment]
- Про принты и гвозди:
Гвозди отодрали, в место принтов теперь нормальный логгер, где это необходимо.  Остались только в тех местах, где это интерфейс для работы в командной строке.

- Про тесты:
#46 До покрытия SQLite нам конечно далеко. Добавили функциональные тесты API и немного тестирования логики в сендерах. Еще есть заготовка для интеграционного тестирования, но пока им у нас занимается команда ручных тестировщиков. А так обязательно будет selenium и автотесты :)

- Про обработку ошибок:
#47 Немного поправили обработку ошибок, два раза письмо не улетит. Вопрос с истории отправки пока остаётся открытым, как возможное решение проблемы видится отдельная очередь, куда будем складывать всё неудачные попытки записи в БД, заносить их туда уже после восстановления соединения.

- Про миграции:
Проблему осознаём, но оставили без изменений.

**Comment by BigDeepBlue** [comment]
Про коммиты:
 - не посмотрел содержимое коммитов. Они действительно довольно атомарны. Но тогда лучше оформлять отдельные PR, чтобы контекст задачи был связан с конкретным коммитом (несколькими-при совместной разработке). А то сейчас выходит, что фичи формально описывает 1 PR, за которым 4 коммита. Сколько и каких коммитов надо откатить, чтобы откатить одну из фич? И вы уверены, что в вашей команде все смогут в атомарные коммиты?) Но в целом да, вопрос достаточно дискуссионный, главное, что вы понимаете проблему и аргументировали свое решение

**Comment by BigDeepBlue** [comment]
Оценил вашу командную работу и ревью PR, здорово, что даже правки замечаний вы разделили между членами команды. Работа принята
